lib/cc/engine/markdownlint.rb

```diff
@@ -7,33 +7,36 @@ module Engine
     class Markdownlint
       CONFIG_FILE = "./.mdlrc".freeze
       EXTENSIONS = %w[.markdown .md].freeze
+      UnexpectedMessageFormat = Class.new(StandardError)
 
-      def initialize(root, engine_config, io)
+      def initialize(root, engine_config, io, err_io)
         @root = root
         @engine_config = engine_config
         @io = io
+        @err_io = err_io
         @contents = {}
       end
 
       def run
         return if include_paths.length == 0
 
-        pid, _, out, err = POSIX::Spawn.popen4("mdl", *mdl_options, *include_paths)
+        child = POSIX::Spawn::Child.new("mdl", *mdl_options, *include_paths)
```

Comment:
@wfleming following up to collapsed https://github.com/codeclimate/codeclimate-markdownlint/pull/16#discussion_r85044363

> This may be pre-existing behavior, but does this engine not error out if the underlying mdl process errors? That seems undesirable. Should we have an issue filed for that?

Correct. I'm not sure how to define "the underlying process errored out". Possible definitions, and their flaws:

1. `mdl` has a non-zero exit code -- no good, because it exits 1 when it finds issues in markdown files
1. `mdl` prints _anything_ to stderr -- feels wrong, because maybe it just prints some messages to stderr in non-fatal contexts
1. `mdl` prints _something_ to stderr which looks like a ruby stack trace (just thought of this one. Feels weird though)

The assumption I have is that if `mdl` errors out, there _may_ be ill-formatted output printed to stdout, which would lead to us raising our new error. If not, it'll just silently fail and find no issues. Not great.

I think it would be handy if `mdl` exited:

1. `0` for no issues
1. `1` for unhandled runtime error
1. `2` (or something else) for some issues

But that isn't the case, currently.

Any other thoughts on how we can do this? It feels like we should have a good technique here and it will be obvious when I wake up.